pilot/pkg/xds/filters/filters.go

```diff
@@ -251,38 +253,41 @@ var (
 							"workload_discovery": map[string]any{},
 						},
 					},
-					"upstream_propagation": []any{
-						map[string]any{
-							"istio_headers": map[string]any{},
-						},
-					},
 				}),
 		},
 	}
-	// TODO https://github.com/istio/istio/issues/46740
-	// false values can be omitted in protobuf, results in diff JSON values between control plane and envoy config dumps
-	// long term fix will be to add the metadata config to istio/api and use that over TypedStruct
-	SidecarOutboundMetadataFilterSkipHeaders = &hcm.HttpFilter{
-		Name: MxFilterName,
+
+	InjectIstioHeaders = &hcm.HttpFilter{
+		Name: "inject-istio-headers",
 		ConfigType: &hcm.HttpFilter_TypedConfig{
-			TypedConfig: protoconv.TypedStructWithFields("type.googleapis.com/io.istio.http.peer_metadata.Config",
-				map[string]any{
-					"upstream_discovery": []any{
-						map[string]any{
-							"istio_headers": map[string]any{},
-						},
-						map[string]any{
-							"workload_discovery": map[string]any{},
+			TypedConfig: protoconv.MessageToAny(&hm.HeaderMutation{
+				Mutations: &hm.Mutations{
+					RequestMutations: []*mutations.HeaderMutation{
+						{
+							Action: &mutations.HeaderMutation_Append{
+								Append: &core.HeaderValueOption{
+									Header: &core.HeaderValue{
+										Key:   "x-envoy-peer-metadata",
+										Value: "%ENVIRONMENT(ISTIO_PEER_METADATA)%",
+									},
+									AppendAction: core.HeaderValueOption_OVERWRITE_IF_EXISTS,
+								},
+							},
 						},
-					},
-					"upstream_propagation": []any{
-						map[string]any{
-							"istio_headers": map[string]any{
-								"skip_external_clusters": true,
```

Comment:
Did we lose this?

Reply:
This gets a lot easier actually because skipping is done per cluster.